design/envoy-shutdown.md

```diff
@@ -0,0 +1,51 @@
+# Envoy Shutdown Probes
+
+Status: _Draft_
+
+This proposal describes how to add better support for identifying the status of Envoy connections during a restart or redeploy of Envoy.
+
+## Goals
+
+- Provide a mechanism to provide feedback on open connections & load inside an Envoy process
+- Allow for a Envoy fleet roll-out minimizing connection loss
+
+## Non Goals
+
+- Guarantee zero connection loss during a roll-out
+
+## Background
+
+The Envoy process, the data path component of Contour, at times needs to be re-deployed.
+This could be due to an upgrade, a change in configuration, or a node-failure forcing a redeployment.
+
+Contour currently implements a `preStop` hook in the container which signals to Envoy to begin draining connections.
+Once Envoy begins draining, the readiness probe on the pod triggers to fail, which in turn causes that instance of Envoy to stop accepting new connections.
+
+The main problem is that the `preStop` hook (which sends the `/healthcheck/fail` request) does not wait until Envoy has drained all connections, so when the container
+is restarted, users can receive errors.
+
+This design looks to add a new component to Contour which will allow for a way to understand if open connections exist in Envoy before sending a `SIGTERM`.
+
+## High-Level Design
+
+Implement a new sub-command of Contour named `envoyshutdown` which will handle sending the healthcheck fail request to Envoy and then begin polling the http/https listeners for active connections from 
+the `/stats` endpoint available on `localhost:9001`.
+
+## Detailed Design
+
+- Implement new subcommand in Contour called `envoyshutdown`
+- This command will expose an http endpoint over a specific port (e.g. `8090`) and path `/shutdown`. 
+- A new container will be added to the Envoy Daemonset pod which will run this new sub-command (e.g. `contour envoyhealthcheck`)
+- This new container as well as the Envoy container will be updated to use an http preStop hook (see example below)
```

Comment:
Example missing?

Also, does the Envoy container need a prestop hook if this one has one? I don't quite understand how that works. Does the Envoy prestop hook hit the new container as well? Or something in the Envoy container?

Either the example or some more explanation would probably help.

Reply:
Ahh right, forgot to add the examples (DOH). 

Since we can't have the hook be in the Envoy container, both containers will need a hook to stay up.

Reply:
Ah, okay. And both the Envoy container and the new container use the same prestophook, calling to the /shutdown endpoint on port 8090 (which is serviced by the new container). This will effectively make the two containers coterminous (although it won't be at exactly the same time because networks handwave handwave). Nice.

Is the idea that the `/healthz` endpoint on the new container will keep track of Envoy and become unhealthy if Envoy does? Otherwise, an Envoy lockup could mean that things get stuck maybe? I'm not sure how this would work, but it can probably be an implementation detail.

Reply:
The `/healthz` endpoint was just on the shutdown container since it's a long-running process and there were concerns that if it was down, the proper shutdown sequence wouldn't happen correctly. 

Do you think this design should account for if Envoy fails somehow? There are a number of scenarios that we'd need to look at, but I think it should be a new issue:

1. **Envoy terminates:** Kubernetes should restart
2. **Envoy stops responding**: The readiness probe existing today should take that Envoy out of service (Needs addressed since there's no automatic remedy for this). A liveliness probe could be used here but would need to be very cautious about how this gets implemented. 
3. **Envoy loses gRPC connection to Contour**: Not handled today

Reply:
I think that I agree with the behaviours @stevesloka enumerates above. Note that envoy doesn't become healthy until it has received an initial configuration from Contour, so in case (3), I would expect Envoy to keep running with its current configuration (and no changes).